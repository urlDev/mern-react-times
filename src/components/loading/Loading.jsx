import DarkLoadingSrc from "../../assets/loadingLogo.svg";

import { LoadingContainer } from "./Loading.styles";

const Loading = ({ height, logoSrc, login }) => {
  return (
    <LoadingContainer height={height} login={login}>
      <div style={{ display: "flex", flexDirection: "column" }}>
        <img src={logoSrc || DarkLoadingSrc} alt="loading logo" />
      </div>
    </LoadingContainer>
  );
};

export default Loading;
