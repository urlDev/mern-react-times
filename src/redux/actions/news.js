import axios from 'axios';

export const FETCH_POPULAR_SUCCESS = 'FETCH_POPULAR_SUCCESS';
export const FETCH_STORY_SUCCESS = 'FETCH_STORY_SUCCESS';
export const FETCH_NEWS_ERROR = 'FETCH_NEWS_ERROR';
export const CHANGE_HEADER = 'CHANGE_HEADER';

export const fetchPopularSuccess = (news) => ({
    type: FETCH_POPULAR_SUCCESS,
    payload: news,
});

export const fetchStorySuccess = (news) => ({
    type: FETCH_STORY_SUCCESS,
    payload: news,
});

export const changeHeader = (header) => ({
    type: CHANGE_HEADER,
    payload: header,
});

export const fetchNewsError = (error) => ({
    type: FETCH_NEWS_ERROR,
    payload: error,
});

export const fetchTopStories = (topStories) => async(dispatch) => {
    try {
        const response = await axios.get(
            `https://api.nytimes.com/svc/topstories/v2/${topStories}.json?api-key=${process.env.REACT_APP_NEWS_API_KEY}`,
        );
        const data = await response.data;
        return dispatch(fetchStorySuccess(data));
    } catch (error) {
        return dispatch(fetchNewsError(error));
    }
};

export const fetchMostPopular = () => async(dispatch) => {
    try {
        const response = await axios.get(
            `https://api.nytimes.com/svc/mostpopular/v2/viewed/7.json?api-key=${process.env.REACT_APP_NEWS_API_KEY}`,
        );
        const data = await response.data;
        return dispatch(fetchPopularSuccess(data));
    } catch (error) {
        return dispatch(fetchNewsError(error));
    }
};