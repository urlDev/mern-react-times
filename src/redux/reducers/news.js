import {
  FETCH_NEWS_ERROR,
  FETCH_POPULAR_SUCCESS,
  FETCH_STORY_SUCCESS,
  CHANGE_HEADER,
  CLEAN_STATE,
  CLEAR_ERROR,
  SET_WIDTH,
  OPEN_RESPONSIVE_MENU,
  CLOSE_RESPONSIVE_MENU,
} from "../actions/news";

const initialState = {
  loadingNews: true,
  popular: [],
  story: [],
  errorNews: null,
  header: "home",
  width: window.innerWidth,
  responsiveMenu: false,
};

const newsReducer = (state = initialState, action) => {
  switch (action.type) {
    case CLEAR_ERROR:
      return {
        ...state,
        errorNews: null,
      };
    case FETCH_POPULAR_SUCCESS:
      return {
        ...state,
        loadingNews: false,
        popular: action.payload.results.slice(0, 4),
      };
    case FETCH_STORY_SUCCESS:
      return {
        ...state,
        loadingNews: false,
        story: action.payload.results.slice(0, 3),
      };
    case FETCH_NEWS_ERROR:
      return {
        ...state,
        errorNews: action.payload,
      };
    case CHANGE_HEADER:
      return {
        ...state,
        header: action.payload,
      };
    case SET_WIDTH:
      return {
        ...state,
        width: action.payload,
      };
    case OPEN_RESPONSIVE_MENU:
      return {
        ...state,
        responsiveMenu: true,
      };
    case CLOSE_RESPONSIVE_MENU:
      return {
        ...state,
        responsiveMenu: false,
      };
    default:
      return state;
  }
};

export default newsReducer;
