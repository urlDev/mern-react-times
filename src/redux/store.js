import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));