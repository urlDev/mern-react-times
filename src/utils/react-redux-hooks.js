// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
