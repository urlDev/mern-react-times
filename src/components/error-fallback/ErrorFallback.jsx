import { useSelector } from "utils/react-redux-hooks";

import { ErrorContainer } from "./ErrorFallback.styles";

const ErrorFallback = ({ height, error }) => {
  const { errorChart } = useSelector((state) => state.chart);
  return (
    <ErrorContainer height={height}>
      <h1>Something went wrong.</h1>
      <p>{error ? error.message : null || errorChart["Error Message"]}</p>
      <p>
        If it's limit reach, do come back some other time. Otherwise, please
        refresh!
      </p>
    </ErrorContainer>
  );
};

export default ErrorFallback;
