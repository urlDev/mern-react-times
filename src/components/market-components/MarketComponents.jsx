import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { ErrorBoundary } from "react-error-boundary";

import {
  changeMarketType,
  changeMarketName,
  fetchHomeChart,
  fetchForex,
} from "../../redux/actions/chart";

import ErrorFallback from "../error-fallback/ErrorFallback";
import Loading from "../loading/Loading";
import HomeMarketCards from "../home-market-cards/HomeMarketCards";

import {
  MarketComponentsContainer,
  MarketHeaderContainer,
  MarketTitle,
  MarketMenu,
} from "./MarketComponents.styles";

const MarketComponents = () => {
  // First, I made market types array, the ones that I want to show in home page
  const marketTypes = ["Indexes", "Crypto", "Forex", "Stocks", "Commodities"];
  const { loadingChart, marketType, marketName } = useSelector(
    (chart) => chart.chart
  );
  const dispatch = useDispatch();

  const fetchMarketTypes = (market) => {
    //  One api endpoint can be used for all market quotes
    //  So I first determine the market type
    //  Then, checking which market type client has clicked on
    //  According to market, I pre-determined tickers/symbols
    //  With a switch function, I dispatch the related market types
    //  desired tickers to redux store.
    //  So with all this, I can use one fetch function for all market kinds,
    //  Therefore, less fetch calls, less code.
    //  More comments as it seems😁
    const Indexes = "%5EGSPC,%5ERUA,%5EDJI,%5ENDX,%5EN225,%5EFTSE";
    const Crypto = "BTCUSD,LTCUSD,XLMUSD,BCNUSD,ETHUSD,ETCUSD";
    const Forex = "EURUSD,USDJPY,GBPUSD,EURGBP,EURJPY,GBPJPY";
    const Stocks = "AAPL,FB,GOOG,TSLA,NFLX,AMZN";
    const Commodities = "ZGUSD,CLUSD,HGUSD,SIUSD,PLUSD,BZUSD";

    switch (market) {
      case "Indexes":
        return [
          dispatch(changeMarketType(Indexes)),
          dispatch(changeMarketName("Indexes")),
        ];
      case "Crypto":
        return [
          dispatch(changeMarketType(Crypto)),
          dispatch(changeMarketName("Crypto")),
        ];
      case "Forex":
        return [
          dispatch(changeMarketType(Forex)),
          dispatch(changeMarketName("Forex")),
        ];
      case "Stocks":
        return [
          dispatch(changeMarketType(Stocks)),
          dispatch(changeMarketName("Stocks")),
        ];
      case "Commodities":
        return [
          dispatch(changeMarketType(Commodities)),
          dispatch(changeMarketName("Commodities")),
        ];
      default:
        return Indexes;
    }
  };

  React.useEffect(() => {
    dispatch(fetchForex(marketType));
    dispatch(fetchHomeChart(marketType));
  }, [dispatch, marketType]);

  return (
    <MarketComponentsContainer>
      <MarketHeaderContainer>
        <MarketTitle>Market Snapshot</MarketTitle>
        {marketTypes.map((market, index) => (
          <MarketMenu
            key={index}
            active={market === marketName}
            type="button"
            onClick={() => {
              fetchMarketTypes(market);
            }}
          >
            {market}
          </MarketMenu>
        ))}
      </MarketHeaderContainer>
      {loadingChart ? (
        <Loading />
      ) : (
        <ErrorBoundary FallbackComponent={ErrorFallback}>
          <HomeMarketCards />
        </ErrorBoundary>
      )}
    </MarketComponentsContainer>
  );
};

export default MarketComponents;
