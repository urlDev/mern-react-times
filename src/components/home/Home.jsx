import React from "react";
import { useSelector, useDispatch } from "utils/react-redux-hooks";
import { Route, Switch } from "react-router-dom";

import {
  fetchTopStories,
  fetchMostPopular,
  changeHeader,
} from "redux/actions/news";

import Loading from "components/loading/Loading";
import ErrorFallback from "components/error-fallback/ErrorFallback";
import StoryComponents from "components/story-components/StoryComponents";
import MarketComponents from "components/market-components/MarketComponents";
import MarketDetails from "components/market-details/MarketDetails";
import ResponsiveNav from "components/responsive-nav/ResponsiveNav";
import SearchResults from "components/search-results/SearchResults";

import { topics } from "components/story-topic-headers/StoryTopicHeaders";

const Home = () => {
  const { header, loadingNews, popular, width } = useSelector(
    (news) => news.news
  );
  const { errorChart } = useSelector((chart) => chart.chart);

  const dispatch = useDispatch();
  const path = window.location.pathname;

  React.useEffect(() => {
    // If client jumps to some route (any of those topics) without clicking to menu bar
    // then the app will show the client the relevant results
    if (
      topics.includes(path.slice(1).charAt(0).toUpperCase() + path.slice(2)) ||
      path === "/"
    ) {
      dispatch(changeHeader(path.slice(1)));
    }

    dispatch(fetchTopStories(header));

    if (popular.length === 0) {
      dispatch(fetchMostPopular());
    }
  }, [dispatch, path, popular.length, header]);

  return (
    <>
      <ResponsiveNav />
      <Switch>
        {loadingNews ? (
          <Loading height={width < 1024 ? "600px" : "400px"} />
        ) : (
          <Route
            path={`/${header.toLowerCase()}`}
            component={StoryComponents}
          />
        )}
        <Route path="/details/:symbol" component={MarketDetails} />
        <Route path="/search/:symbol" component={SearchResults} />
      </Switch>
      {errorChart ? <ErrorFallback /> : <MarketComponents />}
    </>
  );
};

export default Home;
