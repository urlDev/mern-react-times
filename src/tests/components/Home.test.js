import React from "react";
import { shallow } from "enzyme";
import moxios from "moxios";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";

import {
  CHANGE_HEADER,
  fetchTopStories,
  FETCH_STORY_SUCCESS,
} from "redux/actions/news";
import { story } from "tests/fixtures/news";

import Home from "components/home/Home";
import MarketComponents from "components/market-components/MarketComponents";
import Loading from "components/loading/Loading";
import ErrorFallback from "components/error-fallback/ErrorFallback";

describe("Testing Home component", () => {
  let store;
  let wrapper;

  beforeEach(() => {
    window.history.pushState({}, "", "/arts");

    jest.spyOn(React, "useEffect").mockImplementationOnce((f) => f());

    jest
      .spyOn(ReactReduxHooks, "useDispatch")
      .mockImplementation(() => store.dispatch);

    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());

    moxios.install();
  });
  afterEach(() => {
    moxios.uninstall();
  });

  describe("Testing component with MarketComponents", () => {
    beforeEach(() => {
      store = mockStore({
        loadingNews: false,
        errorChart: false,
        popular: [],
        header: "Home",
      });

      wrapper = shallow(<Home store={store} />);
    });

    test("Should render the component with MarketComponents", () => {
      expect(wrapper).toHaveLength(1);
      expect(
        wrapper.containsMatchingElement(<MarketComponents />)
      ).toBeTruthy();
    });
    test("Should match the snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
    test("Should dispatch actions in useEffect", async () => {
      moxios.wait(() => {
        const request = moxios.requests.mostRecent();
        request.respondWith({
          status: 200,
          response: story,
        });
      });

      await store.dispatch(fetchTopStories("Arts"));

      const actions = store.getActions();
      expect(actions).toEqual([
        { type: CHANGE_HEADER, payload: "arts" },
        {
          type: FETCH_STORY_SUCCESS,
          payload: story.results,
        },
      ]);
    });
  });
  describe("Testing the component Error and Loading", () => {
    beforeEach(() => {
      store = mockStore({
        loadingNews: true,
        popular: [],
        errorChart: true,
        header: "Home",
      });

      wrapper = shallow(<Home store={store} />);
    });

    test("Should render error and loading components", () => {
      expect(wrapper).toHaveLength(1);
      expect(wrapper.containsMatchingElement(<Loading />)).toBeTruthy();
      expect(wrapper.containsMatchingElement(<ErrorFallback />)).toBeTruthy();
    });
    test("Should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
