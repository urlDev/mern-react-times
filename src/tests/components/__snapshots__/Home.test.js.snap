// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Home component Testing component with MarketComponents Should match the snapshot 1`] = `
<Fragment>
  <ResponsiveNav />
  <Switch>
    <Route
      component={[Function]}
      path="/home"
    />
    <Route
      component={[Function]}
      path="/details/:symbol"
    />
    <Route
      component={[Function]}
      path="/search/:symbol"
    />
  </Switch>
  <MarketComponents />
</Fragment>
`;

exports[`Testing Home component Testing the component Error and Loading Should match snapshot 1`] = `
<Fragment>
  <ResponsiveNav />
  <Switch>
    <Loading
      height="400px"
    />
    <Route
      component={[Function]}
      path="/details/:symbol"
    />
    <Route
      component={[Function]}
      path="/search/:symbol"
    />
  </Switch>
  <ErrorFallback />
</Fragment>
`;
